### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by ahmadiq**

### CODE REVIEW

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 1](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/1/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 2](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/2/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 3](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/3/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 4](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/4/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 5](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/5/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 6](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/6/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 7](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/7/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 8](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/8/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Yikes! You better fix it before anyone else finds out! [Build 9](https://jenkins.tools.stackator.com/job/Stakater/job/ProxyInjector/job/PR-1/9/) has Failed!

**Comment by stakater-user** [comment]
@ahmadiq Image is available for testing. ``docker pull stakater/proxyinjector:PR-1-10``

File: deployments/kubernetes/chart/proxyinjector/templates/configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    cors-origins:
+    {{- range .corsOrigins }}
+    - {{ . | quote }}
+    {{- end }}
+    cors-methods:
+    {{- range .corsMethods }}
+    - {{ . }}
+    {{- end }}
+    resources:
+    {{- range $key, $value := index .resources 0}}
```
Comment (waseem-h) [inline deployments/kubernetes/chart/proxyinjector/templates/configmap.yaml]: You should use toYaml in place of this

**Comment by stakater-user** [comment]
@ahmadiq Image is available for testing. ``docker pull stakater/proxyinjector:PR-1-11``

File: README.md [hunk 2]
```diff
@@ -1,2 +1,75 @@
...
+    The following arguments can either be added to the proxy injector `config.yaml` in the ConfigMap, or as annotations
+    on the target deployments with a `authproxy.stakater.com/` prefix.
+
+    | Key           | Description                                             |
+    |---------------|---------------------------------------------------------|
+    | listen        | the interface the proxy should be listening on          |
+    | upstream-url  | url for the upstream endpoint you wish to proxy         |
+    | resources     | list of resources to proxy uri, methods, roles          |
+    | client-id     | client id used to authenticate to the oauth service     |
+    | client-secret | client secret used to authenticate to the oauth service |
```
Comment (waseem-h) [inline README.md:46]: What about the rest of the properties? Shouldn't they be documented? Or we can just link keycloak proxy's github repo

File: deployments/kubernetes/chart/proxyinjector/templates/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    group: {{ .Values.proxyinjector.labels.group }}
+    provider: {{ .Values.proxyinjector.labels.provider }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+  name: {{ template "proxyinjector-name" . }}
+data:
+  {{- with .Values.keycloakProxyConfigmap.config }}
+  config.yml: |-
+    client-id: {{ .clientId }}
```
Comment (waseem-h) [inline deployments/kubernetes/chart/proxyinjector/templates/configmap.yaml]: We can just use toYaml here to pick up the whole config file from values.yaml
That way its up to the user to properly add config values and we won't have to manually add a new property whenever it comes in

File: internal/pkg/handler/create.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+type patch struct {
+	Spec Spec1 `json:"spec"`
+}
+
+// Handle processes the newly created resource
+func (r ResourceCreatedHandler) Handle(conf []string) error {
+	if r.Resource == nil {
+		logger.Errorf("Resource creation handler received nil resource")
+	} else {
+		name := callbacks.GetDeploymentName(r.Resource)
```
Comment (waseem-h) [inline internal/pkg/handler/create.go]: The Handle method is doing a lot of stuff. Refactor this into smaller and reusable methods.

File: pkg/kube/resourcemapper.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package kube
+
+import (
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// ResourceMap are resources from where changes are going to be detected
+var ResourceMap = map[string]runtime.Object{
+	"deployments": &v1beta1.Deployment{},
```
Comment (waseem-h) [inline pkg/kube/resourcemapper.go]: You're watching only deployments. Watch other resources too

File: internal/pkg/handler/create.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+		namespace := callbacks.GetDeploymentNamespace(r.Resource)
```
Comment (waseem-h) [inline internal/pkg/handler/create.go]: You're assuming that the resource can only be a deployment. It should cater all resources like daemonsets and statefulsets.

File: internal/pkg/handler/create.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+			containerArgs := conf
+
+			for _, arg := range constants.KeycloakArgs {
+				if annotations[constants.AnnotationPrefix+arg] != "" {
+					containerArgs = append(containerArgs, "--"+arg+"="+annotations[constants.AnnotationPrefix+arg])
+				}
+			}
+
+			payload := patch{
```
Comment (waseem-h) [inline internal/pkg/handler/create.go]: Refactor object creations to separate methods

File: internal/pkg/util/interface.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
+package util
```
Comment (waseem-h) [inline internal/pkg/util/interface.go]: This whole file is unused. Please remove if not needed

**Review by waseem-h — CHANGES_REQUESTED** [review]

File: pkg/kube/resourcemapper.go (hunk 5 above)
Comment (ahmadiq) [reply pkg/kube/resourcemapper.go]: getting some api related issues while handling daemonsets and statefulsets. Will resolve and handle in next version.

**Comment by stakater-user** [comment]
@ahmadiq Image is available for testing. ``docker pull stakater/proxyinjector:PR-1-13``

**Pull request merged**